core/crypto/kem/adapter/kem.go

```diff
@@ -0,0 +1,264 @@
+}
+
+// PrivateKey is an adapter for nike.PrivateKey to kem.PrivateKey.
+type PrivateKey struct {
+	privateKey nike.PrivateKey
+	scheme     *Scheme
+}
+
+func (p *PrivateKey) Scheme() kem.Scheme {
+	return p.scheme
+}
+
+func (p *PrivateKey) MarshalBinary() ([]byte, error) {
+	return p.privateKey.MarshalBinary()
+}
+
+func (p *PrivateKey) Equal(privkey kem.PrivateKey) bool {
+	if privkey.(*PrivateKey).scheme != p.scheme {
+		return false
+	}
+	return hmac.Equal(privkey.(*PrivateKey).privateKey.Bytes(), p.privateKey.Bytes())
+}
+
+func (p *PrivateKey) Public() kem.PublicKey {
+	return &PublicKey{
+		publicKey: p.privateKey.Public(),
+		scheme:    p.scheme,
+	}
+}
+
+// Scheme is an adapter for nike.Scheme to kem.Scheme.
+// See docs/specs/kemsphinx.rst for some design notes
+// on this NIKE to KEM adapter.
+type Scheme struct {
+	nike nike.Scheme
+}
+
+var _ kem.Scheme = (*Scheme)(nil)
+var _ kem.PublicKey = (*PublicKey)(nil)
+var _ kem.PrivateKey = (*PrivateKey)(nil)
+
+// FromNIKE creates a new KEM adapter Scheme
+// using the given NIKE Scheme.
+func FromNIKE(nike nike.Scheme) *Scheme {
+	return &Scheme{
+		nike: nike,
+	}
+}
+
+// Name of the scheme
+func (a *Scheme) Name() string {
+	return a.nike.Name()
+}
+
+// GenerateKeyPair creates a new key pair.
+func (a *Scheme) GenerateKeyPair() (kem.PublicKey, kem.PrivateKey, error) {
+	pubkey, privkey, err := a.nike.GenerateKeyPair()
+	if err != nil {
+		return nil, nil, err
+	}
+	return &PublicKey{
+			publicKey: pubkey,
+			scheme:    a,
+		}, &PrivateKey{
+			privateKey: privkey,
+			scheme:     a,
+		}, nil
+}
+
+// Encapsulate generates a shared key ss for the public key and
+// encapsulates it into a ciphertext ct.
+func (a *Scheme) Encapsulate(pk kem.PublicKey) (ct, ss []byte, err error) {
+	seed := make([]byte, a.EncapsulationSeedSize())
+	_, err = rand.Reader.Read(seed)
+	if err != nil {
+		return
+	}
+	return a.EncapsulateDeterministically(pk, seed)
+}
+
+func hashSharedSecretWithPublicKeys(publicKeySize int, ss []byte, pubkey1 []byte, pubkey2 []byte) []byte {
+	var h blake2b.XOF
+	var err error
+	if len(ss) != 32 {
+		sum := blake2b.Sum256(ss)
+		h, err = blake2b.NewXOF(uint32(publicKeySize), sum[:])
+	} else {
+		h, err = blake2b.NewXOF(uint32(publicKeySize), ss)
+	}
+	if err != nil {
+		panic(err)
+	}
+	_, err = h.Write(pubkey1)
+	if err != nil {
+		panic(err)
+	}
+	_, err = h.Write(pubkey2)
+	if err != nil {
+		panic(err)
+	}
+	ss2 := make([]byte, len(ss))
+	_, err = h.Read(ss2)
+	if err != nil {
+		panic(err)
+	}
+	return ss2
+}
+
+// Returns the shared key encapsulated in ciphertext ct for the
+// private key sk.
+// Implements DECAPSULATE as described in NIKE to KEM adapter,
+// see docs/specs/kemsphinx.rst
+func (a *Scheme) Decapsulate(sk kem.PrivateKey, ct []byte) ([]byte, error) {
+	if len(ct) != a.CiphertextSize() {
+		return nil, kem.ErrCiphertextSize
+	}
+	pk, err := a.UnmarshalBinaryPublicKey(ct)
```

Comment:
The naming scheme where `sk` belongs to `a` and `pk` belongs to the other party is confusing.